src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/LinkedList.cs

```diff
@@ -0,0 +1,192 @@
+                return;
+            }
+
+            Debug.Assert(_first is not null);
+            Debug.Assert(_last is not null);
+
+            _last!.Next = newNode;
+            _last = newNode;
+        }
+
+        public void Add(T value)
+        {
+            LinkedListNode<T> newNode = new LinkedListNode<T>(value);
+
+            lock (this)
+            {
+                UnsafeAdd(newNode);
+            }
+        }
+
+        public bool AddIfNotExist(T value, Func<T, T, bool> compare)
+        {
+            lock (this)
+            {
+                LinkedListNode<T>? current = _first;
+                while (current is not null)
+                {
+                    if (compare(value, current.Value))
+                    {
+                        return false;
+                    }
+
+                    current = current.Next;
+                }
+
+                LinkedListNode<T> newNode = new LinkedListNode<T>(value);
+                UnsafeAdd(newNode);
+
+                return true;
+            }
+        }
+
+        public T? Remove(T value, Func<T, T, bool> compare)
+        {
+            lock (this)
+            {
+                LinkedListNode<T>? previous = _first;
+                if (previous is null)
+                {
+                    return default;
+                }
+
+                if (compare(previous.Value, value))
+                {
+                    _first = previous.Next;
+                    if (_first is null)
+                    {
+                        _last = null;
+                    }
+                    return previous.Value;
+                }
+
+                LinkedListNode<T>? current = previous.Next;
+
+                while (current is not null)
+                {
+                    if (compare(current.Value, value))
+                    {
+                        previous.Next = current.Next;
+                        if (object.ReferenceEquals(_last, current))
+                        {
+                            _last = previous;
+                        }
+
+                        return current.Value;
+                    }
+
+                    previous = current;
+                    current = current.Next;
+                }
+
+                return default;
+            }
+        }
+
+        public void AddFront(T value)
+        {
+            LinkedListNode<T> newNode = new LinkedListNode<T>(value);
+
+            lock (this)
+            {
+                newNode.Next = _first;
+                _first = newNode;
+            }
+        }
+
+        // Note: Some consumers use this GetEnumerator dynamically to avoid allocations.
+        public Enumerator<T> GetEnumerator() => new Enumerator<T>(_first);
+        IEnumerator<T> IEnumerable<T>.GetEnumerator() => GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+    }
+
+    // Note: Some consumers use this Enumerator dynamically to avoid allocations.
```

Comment:
What does "dynamically" mean here?

Reply:
I think this is originally introduced to address the ask in the comment https://github.com/dotnet/runtime/pull/40544/files#r467197582

Reply:
Ok... what does it mean? :-)

Reply:
My understanding is:

Originally we had 

```C#
public Enumerator<KeyValuePair<string, object?>> GetEnumerator() => new Enumerator<KeyValuePair<string, object?>>(_first); 
```

which was not reachable by the public APIs and was causing the linker to trim it. The proposed fix was to add the attribute `[DynamicDependency(DynamicallyAccessedMemberTypes.PublicMethods, typeof(TagsLinkedList))]` to this method to stop the linker from trimming it. @eerhardt suggested a better solution to change the enumerator implementation to something like:

```
        public Enumerator<T> GetEnumerator() => new Enumerator<T>(_first);
        IEnumerator<T> IEnumerable<T>.GetEnumerator() => GetEnumerator();
        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

```

So, the comment describing that. I guess the word `dynamically` is picked up from the attribute name `DynamicDependency`.

CC @CodeBlanch if he can clarify more.

Reply:
What are the consumers that are accesing this field dynamically (I assume that it means reflection)?

The burden is on the consumers to ensure that the method is not trimmed when they access it via reflection.

Reply:
> The burden is on the consumers 

IIRC, the issue was the linker was removing the non-public, unused method during our build of the library. And so the method wasn’t even ending up in our shipped assembly.

I question whether this scenario is worthwhile, since calling the method through reflection is going to box the struct anyway as an ‘object’ return. But I remember someone had reasoning for doing this.

Reply:
In OpenTelemetry we build a DynamicMethod that will access the struct enumerator and loop on it without boxing: https://github.com/open-telemetry/opentelemetry-dotnet/blob/a0fcbc0a35ad63ff21f1efcda45491570841985b/src/OpenTelemetry.Api/Internal/EnumerationHelper.cs#L79

Usage example: https://github.com/open-telemetry/opentelemetry-dotnet/blob/a0fcbc0a35ad63ff21f1efcda45491570841985b/src/OpenTelemetry.Api/Internal/ActivityHelperExtensions.cs#L75

Reply:
It is a bad practice to access framework private implementation details using reflection. Our default stance is that it is not supported scenario. When we break library does access private implementation details using reflection, we consider it by design.

Could you please delete the `BindingFlags.NonPublic` from 
https://github.com/open-telemetry/opentelemetry-dotnet/blob/a0fcbc0a35ad63ff21f1efcda45491570841985b/src/OpenTelemetry.Api/Internal/EnumerationHelper.cs#L167

If you open telemetry needs to access this enumerator method, we should make it official public API instead.